docs/keploy.1
\" Process this file with
.\" groff -man -Tascii keploy.1
.\"
.TH keploy 1 "MARCH 2010" Linux "User Manuals"
.SH NAME
keploy \- A cli ssh public key deployment utility. 
.SH SYNOPSIS
keploy [ -Ahkrqvy ] [ -o
.I OLD_ID_FILE
.B ] [ -f
.I TARGETS_FILE
.B ] [ -i
.I ID_FILE
.B ] [ -l
.I LOGIN_NAME
.B ] [ -t
.I TIMEOUT
.B ]
.SH DESCRIPTION
.B keploy
is a python application that allows you to deploy your ssh
public key to remote systems without having to remember all the
little things, like file permissions.

.SH OPTIONS
.IP -A
Toggle agent forwarding on/off for remote(s)
.IP "-o OLD_ID_FILE"
Specifies which old identity to replace with new idenity
.IP "-f TARGET_FILE"
Read list of targets from a text file, with one line per host
.IP "-i ID_FILE"
Selects a file from which the public identity will be read, default is
.I ~/.ssh/identity.pub
for protocol version 1, and
.I ~/.ssh/id_rsa.pub
and
.I ~/.ssh/id_dsa.pub
for protocol version 2
.IP -h
Show the help message and exit
.IP -k
Read list of targets from
.I ~/.ssh/known_hosts
files
.IP "-l LOGIN_NAME"
Specifies the user to log in as on the remote machine(s)
.IP -r
Remove primary identity from remote machine(s)
.IP -q
Quiet the output
.IP -t
Set the SSH ConnectTimeout value in seconds, see ssh_config (5) for details
.IP -v
Give verbose output
.IP -y
Accept and save new host key if visiting host for first time
.SH BUGS
no known bugs
.SH AUTHOR
Greg Swift <gregswift@gmail.com>
.SH "SEE ALSO"
.BR ssh (1),
.BR ssh-add (1),
.BR ssh-agent (1),
.BR ssh-keygen (1)



CHANGELOG
Release 0.7 - Nov 8th, 2012 - Greg Swift
- Added timeout to cli options
- Added autoaccept of unknown SSH public keys as cli option

Release 0.6 - Jun 22nd, 2010 - Greg Swift 
- Re-wrote the entire application to be module based
- Changed cli switch for old identity file from -c to -o
- Restructured project to utilize distutils
- Re-wrote layout of changelog file

Release 0.5 - Feb 4th, 2010 - Greg Swift
- Fix bug #1 - keploy properly error on inaccessible id file
- Added over_ride default value in getIdentity function
- Added CHANGELOG file

Release 0.4 - Mar 06, 2008 - Greg Swift
- First proper release of the code


keploy/keploy.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# keploy - SSH Public Key Deplyment Utility
# Copyright © 2007 Greg Swift gregswift@gmail.com
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os
import signal
import sys

"""
 Define Variables
"""
DEBUG = True
VERSION = "0.5"

PW_WARN = '\nNOTICE: You may be prompted for you password,\n'
PW_WARN += 'NOTICE: this is directly from the ssh client, not keploy\n'

class KeployVariables:
  def __init__(self, user=None):
    self.ssh_bin = findSSHbin()
    self.ssh_config = '/etc/ssh/ssh_config'
    if (not os.access(self.ssh_config, os.R_OK)):
      self.ssh_config = None
    self.ssh_home_dir = os.path.expanduser('~/.ssh')
    self.tmp_file = os.path.join(self.ssh_home_dir, 'keyploy.tmp')
    self.id_files = (os.path.join(self.ssh_home_dir, 'id_rsa.pub'),
                     os.path.join(self.ssh_home_dir,'id_dsa.pub'),
                     os.path.join(self.ssh_home_dir,'identity.pub'))
    self.host_files = [os.path.join(self.ssh_home_dir,'known_hosts'),
                       os.path.join(self.ssh_home_dir,'known_hosts2')]
    self.auth_keys_file = os.path.join(self.ssh_home_dir,'authorized_keys')
    self.user = user

  def set(self, name, value):
    setattr(self, name, value)

"""
 Define Output Functions and cleanUp Function
"""
def standardOut(msg, on=True):
  """When quiet is not enabled, display msg provided """
  if (on):
    print msg

def debugOut(msg, name=None, on=DEBUG):
  """When debug is enabled, output is printed to the display."""
  if (name is not None):
    name = name+': '
  else:
    name = ''
  try:
    if (on):
      print 'DEBUG: %s%s' % (name, msg)
    else:
      pass
  except:
    print 'MALFORMED DEBUG: %s%s' % (name, msg)

def errorOut(msg, ret=1):
  """When debug is enabled, output is printed to the display.

 returns nothing"""
  type = ['WARNING', 'ERROR']
  out = '%s: %s' % (type[ret], msg)
  if (ret == 0):
    print out
  elif (ret == 1):
    out += '\n'
    sys.stderr.write(out)
    cleanUp(ret)

def cleanUp(ret):
  """Do a clean and proper sys.exit()

 executes sys.exit(int(ret))"""
  sys.exit(ret)


"""
 Define Primary Functions
"""
def getSSHbinary():
  """
  Try to locate the ssh binary using which.
  If 'None' we found nothing
  If 'False' it is not executeable (shouldn't happen, but why not test?)

  returns str(ssh_bin)
	"""
  ssh_bin = os.popen('which ssh').read().strip()
  if (ssh_bin == ''):
    ssh_bin = None
  elif (os.access(ssh_bin, os.X_OK)):
    ssh_bin = False
  return ssh_bin

def getHostsFromFile(get_from=None, known=False, verbose=True):
  """
 Read hosts from provided file.  If no file specified,
 try to read ~/.ssh/known_hosts

 returns list(hosts)
 """
  debugOut('getHostsFromFile(get_from=%s, known=%s, verbose=%s)' % (get_from,
      known, verbose))
  global host_files
  if (get_from is not None):
    if (os.access(get_from, os.R_OK)):
      standardOut('Using user-defined host list from file: %s' % (
        get_from), verbose)
      host_files = [get_from]
  else:    
    if (known):
      # First check to see if we can even get useful data out of the 
      # known_hosts file
      execute = "grep HashKnownHosts %s | awk '{print $2}'" % (ssh_config)
      debugOut(execute, '\tExecuting')
      is_hash = os.popen(execute).read().strip()
      debugOut(is_hash, '\tIs known_hosts hashed')
      if (is_hash == "yes"):
        msg = "The known_hosts files are hashed.  Please "
        msg += "specify host or alternate file to parse at cli"
        errorOut(msg)
      for f in host_files:
        try:
          os.path.exists(f)
        except:
          pass
        else:
          host_files.remove(f)
  hosts = []
  for x in host_files:
    execute = 'cat %s | cut -f1 -d" " | sed -e "s/,.*//g" | uniq' % (x)
    debugOut(execute, '\tExecuting')
    si, so, se = os.popen3(execute)
    for h in so.readlines():
      h = h.strip()
      debugOut(h, '\tFound Host')
      hosts.append(h)
  return list(hosts)

def getIdentity(id_file=None, verbose=True):
  """Read the identity file into a variable so that it is easier to push
to the external host(s)

 returns str(identity)"""
  debugOut('getIdentity(id_file=%s, verbose=%s)' % (id_file, verbose))
  global id_files
  if (id_file is not None):
    debugOut(id_file, '\tOver-riding id_files with')
    over_ride = True
    id_files = [os.path.abspath(os.path.expanduser(id_file))]
    standardOut('\tReading identity from:\n\t\t%s' % (id_file), verbose)
  for id_file in id_files:
    debugOut(id_file, '\tTrying to grab identity from')
    if (os.access(id_file, os.R_OK)):
      identity = os.popen('head -1 %s' % (id_file)).read().strip()
      msg = ''
      break
    else:
      if (over_ride):
        msg = 'Could not find/access specified file, %s' % (id_file)
      else:
        msg = 'Could not find/access default identity files'
  if (msg is ''):
    standardOut('\tFound identity:\n\t\t%s' % (id_file), verbose)
    debugOut(identity, '\tUsing identity')
  else:
    errorOut(msg)
  return str(identity)

def toggleAgentForwarding(on, ssh_call, end_ssh_call, verbose=True):
  """
 This function 
 """
  command = ''
  forward_option = 'ForwardAgent'
  ssh_user_config_file = os.path.join(ssh_home_dir, 'config')
  command +=  'grep -v \"%s\" %s > %s 2> /dev/null;' % (
    forward_option, ssh_user_config_file, tmp_file)
  command += 'mv -f %s %s 2>/dev/null; chmod 600 %s 2>/dev/null;' % (
    tmp_file, ssh_user_config_file, ssh_user_config_file)
  if (on):
    command += 'echo \'%s yes\' >> %s 2> /dev/null; grep \"%s\" %s' % (
      forward_option, ssh_user_config_file, forward_option, ssh_user_config_file)
  remote_command=ssh_call+command+end_ssh_call
  debugOut(remote_command, 'Executing')
  ret = os.popen(remote_command).readlines()
  if (ret):
    if (on):
      status = (True, 'enabled')
    else:
      status = (False, 'failed to disable')
    for line in ret:
      debugOut(line)
  else:
    if (on):
      status = (False, 'failed to enable')
    else:
      status = (True, 'disabled')
  return status
  #standardOut('\t\tAgent Forwarding: %s' % (status), verbose)

def pushToRemote(vars, options, hosts):
  for host in hosts:
    ssh_call = '%s -qq %s %s \'' % (vars.ssh_bin, options.userid, host)
    end_ssh_call = '\''
    standardOut('\tWorking on host: %s' % (host),
                options.verbose)
    if (options.forward):
      toggleAgentForwarding(False, ssh_call, end_ssh_call, options.verbose)
    command = ''
    try:
      old_identity
    except:
      use_identity = identity
    else:
      use_identity = old_identity
    # Make ssh home dir and set permissions
    command += 'mkdir %s &> /dev/null; chmod 700 %s;' % (
      ssh_home_dir, ssh_home_dir)
    # Remove identity from file and set permissions
    command +=  'grep -v \"%s\" %s > %s 2> /dev/null;' % (
      identity, auth_keys_file, tmp_file)
    command += 'mv -f %s %s 2>/dev/null;' % (tmp_file, auth_keys_file)
    command += 'chmod 600 %s 2> /dev/null;' % (auth_keys_file)
    if (not options.remove or options.old_id_file):
      use_identity = identity
      # Then append the pub identity into the authorized_keys file, and grep it
      command += 'echo \'%s\' >> %s 2> /dev/null; chmod 600 %s; grep \"%s\" %s' % (
        use_identity, auth_keys_file, auth_keys_file, use_identity, auth_keys_file)
    remote_command=ssh_call+command+end_ssh_call
    debugOut(remote_command, 'Executing')
    ret = os.popen(remote_command).readlines()
    if (ret):
      if (options.old_id_file):
        status = 'changed'
      elif (options.remove):
        status = 'failed to remove'
      else:
        status = 'deployed'
      for line in ret:
        debugOut(line)
    else:
      if (options.remove):
        status = 'removed'
      else:
        status = 'failed to deploy'
    standardOut('\t\tPublic Identity Key: %s' % (status), options.verbose)

    if (options.forward):
      if (not options.remove or options.old_id_file):
        toggleAgentForwarding(True, ssh_call, end_ssh_call, options.verbose)
